Hack: Force UIEdgeInsets.zero to always come from the SDK

jrose-apple opened the pull request:
...instead of relying on the one in the overlay in pre-4.2 versions of Swift. This caused crashes in deserialization, which (deliberately) doesn't respect availability.

There are three changes here:

- Remove `UIEdgeInsets.zero` and `UIOffset.zero` from the UIKit overlay.
- Always use the 4.2 name for `UIEdgeInsetsZero` and `UIOffsetZero` from the underlying UIKit framework. (This is the nested name.)
- Ignore the unavailability messages for those two constants in pre-4.2 Swift, since we're now relying on them being present.

The latter two, the compiler changes, can go away once UIKit's API notes no longer specify different pre-4.2 behavior, but meanwhile we need to keep compatibility with the SDKs released in Xcode 10b1.

[SR-7879](https://bugs.swift.org/browse/SR-7879) / rdar://problem/40735990

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
Let's try that again.

@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
@swift-ci Please test source compatibility

slavapestov commented on validation-test/Serialization/SR7879.swift:
```diff
@@ -0,0 +1,10 @@
+
+import UIKit
+
+_ = UIEdgeInsets.zero
```
Can you try defining default arguments with these values as well so that we can test serializing references to these globals too?

jrose-apple replied:
Good call. The original bug showed up just a mention of the SIL global in the partial swiftmodules, but having an actual reference to it is important.

slavapestov commented:
@jrose-apple I'm curious why XREFs to imported types don't respect the module filter. This would have fixed the problem without adding a special case.

jrose-apple commented:
That doesn't actually fix the problem, because we don't want to find the one in the overlay. We want the one in the actual Clang module.

(The original motivation, though, was when we thought swiftmodule was going to be a stable format, in which case it needs to cope with declarations being moved from one framework to another, something that's legal in Objective-C.)

jrose-apple committed:
Hack: Force UIEdgeInsets.zero to always come from the SDK

...instead of relying on the one in the overlay in pre-4.2 versions of
Swift. This caused crashes in deserialization, which (deliberately)
doesn't respect availability.

There are three changes here:

- Remove UIEdgeInsets.zero and UIOffset.zero from the UIKit overlay.
- Always use the 4.2 name for UIEdgeInsetsZero and UIOffsetZero from
  the underlying UIKit framework. (This is the nested name.)
- Ignore the unavailability messages for those two constants in
  pre-4.2 Swift, since we're now relying on them being present.

The latter two, the compiler changes, can go away once UIKit's API
notes no longer specify different pre-4.2 behavior, but meanwhile we
need to keep compatibility with the SDKs released in Xcode 10b1.

https://bugs.swift.org/browse/SR-7879

jrose-apple commented:
@swift-ci Please smoke test Linux
